[TG-1403] Correctly specialise classes with generic arrays or generic types

thk123 opened the pull request:
Depends on: #1555 :white_check_mark: 

Previous specialisation code only dealt with types that were directly `T`, this extends this to recursively check arrays (`T []`) and generic types (`Generic<T>`) which can be specialised.

Commit:
Correctly handle generic classes that have a array field

Previously we only converted types that were precisely a generic
parameter. Now we deal with the case where the array is T type'd.

Commit:
Renaming to_java_generic_class_type to remove spurious s

Commit:
Unit utility for symbol types

Commit:
Extend the specialisation code to handle generic fields

Previously we didn't handle generic fields when we were doing
specialisation, meaning they were missed. This resolves this and adds a
unit test to demonstrate it.

NlightNFotis commented on unit/testing-utils/require_type.cpp:
```diff
@@ -289,3 +289,20 @@ require_type::require_java_non_generic_class(const typet &class_type)
+/// \param type: The type to check
+/// \param identifier: The identifier the symbol type should have
+/// \return The cast version of the input type
+const symbol_typet &require_type::require_symbol(const typet &type,
```
This function looks very useful. Nicely done.

NlightNFotis commented on unit/testing-utils/require_type.cpp:
```diff
@@ -289,3 +289,20 @@ require_type::require_java_non_generic_class(const typet &class_type)
+    REQUIRE(result.get_identifier()==identifier);
+  }
+  return result;
+
```
Nit: can this blank line be deleted?

NlightNFotis reviewed (approved):
This looks solid. Nicely done.

Commit:
Extending tests to deal with specialising with arrays when array fields

thk123 commented:
Linter errors are related to lambdas

mgudemann commented on src/java_bytecode/generate_java_generic_type.cpp:
```diff
@@ -98,6 +80,103 @@ symbolt generate_java_generic_typet::operator()(
+
+    INVARIANT(results.has_value(), "generic component type not found");
+
+    if(results)
```
isn't this tested in the above invariant using `.has_value()`?

mgudemann reviewed (approved):
LGTM, modulo the potential simplification

Commit:
Enforce condition that generic references must refer to generic classes

We require that all generic references refer to generic classes

Also pulled out the conversion out of the loop and gave the variable a
more logical name

Added check for generic utils function

Commit:
Removed redundant if statement

thk123 commented:
Thanks for reviews - comments and formatting addressed, will create a TG pointer bump then this should be good to go.
